Merge JS2 to Main

pretzL opened the pull request.

pretzL committed:
initial

pretzL committed:
Merge pull request #1 from pretzL/initial-files

Initial file push

pretzL committed:
Initial API implementation complete,
Added base fetch for Posts,
separated files which could (should?) be hidden later.

pretzL committed:
Added query params to get author, comment and reaction information

pretzL committed:
Extra file prep. Added a components folder for JS, prepared error handling file. Added a barrel file and linked to head. Changed placeholder text for email input on login.html.

pretzL committed:
Created getPosts function which displays returned info in the designated container. Added comments to show what needs to be done still.

pretzL committed:
Removed old hard-coded post HTML

pretzL committed:
Added gap variable

pretzL committed:
Added spacing to post list

pretzL committed:
Add error message code

Sindre1993 committed:
Merge pull request #2 from pretzL/initial-api-implementation

Initial API implementation complete,

pretzL committed:
Changed image folder structure to match existing policy.
Added fallback (default) values for returned API JSON for getPost.
Added container for error handling display.

pretzL committed:
Fixed default value for post media not working

pretzL committed:
Refactored import

pretzL committed:
Filtering out more test posts by other users

pretzL committed:
Changed file extensions to ES6 .mjs

pretzL committed:
Changed script file structure slightly for better organization.
Added register user functionality. HTML page is missing so it isn't connected yet, but will be with page HTML branch.

pretzL committed:
Added values for banner and avatar

pretzL committed:
Created working login authentication which returns the user info payload

pretzL committed:
Added error handling to login auth

Sindre1993 committed:
Merge pull request #22 from pretzL/get-posts

Get posts

Sindre1993 committed:
Merge pull request #23 from pretzL/user-register

User register

Sindre1993 committed:
Added Register.html & added redirect from Sign Up CTA

Added Register.html from Login.html as a base foundation, changed the sign up button on login to redirect the user to the registration form.

pretzL committed:
Added localStorage entry for JWT token.
Added token handling and usage.

Sindre1993 committed:
Updated Register.html

Added input for Avatar and Banner on the registration form on register.html with no required input as we have a fallback for default avatar/banner.

Sindre1993 committed:
Added button to register.html

Added a "Back to Login" button on the registration form.

pretzL committed:
Removed login form action and instead handled that through location redirect in JS.
Added checks to see if the user is logged in. If the user is, allow the user to roam free. If the user is not logged in, they must log in before being able to view the site.

pretzL committed:
Merge pull request #27 from pretzL/Register

Created Register Form

pretzL committed:
Merge branch 'js2' into user-login

Sindre1993 committed:
Updated Profile.html, added banner & profile image along with "follow" button, removed profile image and stats from left sidebar, and made it responsive.

pretzL committed:
Merge pull request #26 from pretzL/user-login

User login

Sindre1993 committed:
Merge pull request #28 from pretzL/is-user-logged-in

Is user logged in

Sindre1993 committed:
Added Error CSS

pretzL committed:
Merge branch 'js2' into profile/banner

pretzL committed:
Merge pull request #30 from pretzL/profile/banner

Profile/banner

pretzL committed:
Updated user side panel to display API data.
Updated create post section to display API data.
Added success message CSS and functionality.
On user login, user info is now saved in localStorage to be used site-wide.
Reorganized JavaScript folder structure in order to support barrel-filing per page.
Connected Register form with JS.

pretzL committed:
Added default value for user profile images

pretzL committed:
Added API data to profile.
Added fallback value for user banner.

pretzL committed:
Consolidated files.

pretzL committed:
Added working post creation, with inputs. Will revise visuals later.

pretzL committed:
Refactored barrel-file system. Each page now has its own mjs file for importing into instead of one for all.
Fixed minor issues with setTimeout functions syntax

pretzL committed:
Changed isUserLoggedIn position in file hierarchy

pretzL committed:
Changed script.mjs file to include only scrips which should be loaded on all pages, and re-added scripts.mjs to all pages

pretzL committed:
Added required RegEx for validating emails.

pretzL committed:
Co-authored-by: Oliver Dipple <Fermain@users.noreply.github.com>

pretzL committed:
Remove vite

pretzL committed:
Time since now works perfectly thanks to dayjs!

pretzL committed:
Added comments

Sindre1993 committed:
Merge pull request #31 from pretzL/user-panel

Updated user side panel to display API data.

Sindre1993 committed:
Merge pull request #33 from pretzL/create-post

Create post

Sindre1993 committed:
Merge pull request #38 from pretzL/email-validation

Email validation

Sindre1993 committed:
Merge pull request #40 from pretzL/time-ago

Time ago

pretzL committed:
Added logout function

pretzL committed:
Fix issue with all posts having a very large invisible section for an image

Sindre1993 committed:
Merge pull request #42 from pretzL/log-out

Added logout function

Sindre1993 committed:
Added Drop Down Menus

Added dropdown menu in navigation for logout, and a drop down for filtering posts on homescreen.

pretzL committed:
Merge pull request #43 from pretzL/sort-by-filter

Added Drop Down Menus

pretzL committed:
Refactored image handling code

pretzL committed:
Refactored image handling into factory functions

pretzL committed:
Removed excess console.logs

Sindre1993 committed:
Created comment section HTML

Comment Section made in HTML

pretzL committed:
Minor code adjustments.
Made inputs on post create 100% width.

pretzL committed:
Merge branch 'js2' into comment-section

pretzL committed:
Merge pull request #45 from pretzL/comment-section

Comment section

pretzL committed:
Merge branch 'js2' into refactoring/fixing

pretzL committed:
Merge pull request #44 from pretzL/refactoring/fixing

Refactoring/fixing

Sindre1993 committed:
Updated Comment Section

Added comment section & buttons for editing and deleting comments/posts.

pretzL committed:
Revert "Refactoring/fixing"

pretzL committed:
Merge pull request #46 from pretzL/revert-44-refactoring/fixing

Revert "Refactoring/fixing"

pretzL committed:
Repeat previous merge, from refactoring/fixing

pretzL committed:
Fix inputs

Sindre1993 committed:
Merge pull request #47 from pretzL/repeat-merge

Repeat previous merge, from refactoring/fixing

pretzL committed:
Created clickable filter functions for author, title, time, today.
Still missing reversable functions and wekk, month etc. filters

pretzL committed:
Added regEx image handling to filter out images which aren't URLs

pretzL committed:
Filter posts created this week

pretzL committed:
Added filter to display posts posted this month

pretzL committed:
Created my posts filter, fixed posts not showing on profile anymore. Will fix so that only own posts show on profile later.

pretzL committed:
Fixed today filter not showing descending

pretzL committed:
Filters now reverse on second click, but seems imp to make the icon change?

pretzL committed:
Removed icons for categories which don't reverse.
Added some spacing between icons and text

pretzL committed:
Initial braindump comment about what needs to be done

pretzL committed:
Rebased one of the getPosts fetch functions to better support filtering in the future

pretzL committed:
Forgot a comment

Sindre1993 committed:
Merge pull request #50 from pretzL/filter-functions

Filter functions

Sindre1993 committed:
Added cogwheel my posts

Added cogwheel to posts, issue persists, cogwheel being added to every post.

pretzL committed:
Cogwheel is now added only to users' own posts.

pretzL committed:
Profiles are now built dynamically if a profile icon is clicked.

pretzL committed:
Merge branch 'js2' into user-profiles

pretzL committed:
Removed comment

pretzL committed:
Added option to add profile URL params

pretzL committed:
Add missing cogwheel to login page

pretzL committed:
Added fallback image handling on registration

pretzL committed:
Merge pull request #54 from pretzL/update-posts

Update posts

pretzL committed:
Merge branch 'js2' into user-profiles

Sindre1993 committed:
Merge pull request #53 from pretzL/user-profiles

User profiles

Sindre1993 committed:
Updated edit post

pretzL committed:
Grabbing the correct post for the edit function now works

pretzL committed:
Edit post now turns body field into input.

Sindre1993 committed:
Added functionality to edit button

Clicking edit changes post-content to inputs.

Sindre1993 committed:
updated Edit Post

pretzL committed:
Attempt to change cards into a form when editing. Not successful yet

Sindre1993 committed:
Created Delete Post Function

Added delete post function to the delete button in the cogwheel of our own posts.

Making confirmation buttons for deleting posts remains.

Sindre1993 committed:
Updated comments

Sindre1993 committed:
Updated comments

pretzL committed:
Edit post system now works fully, with input changing and everything

pretzL committed:
Fixed createPost API response crying about bad image URL format when submitting empty string

Sindre1993 committed:
Merge pull request #55 from pretzL/edit-post

Edit post

pretzL committed:
Merge branch 'js2' into deletePost

pretzL committed:
Remove unnecessary commented out setTimeout

pretzL committed:
Merge pull request #58 from pretzL/deletePost

Delete post

Sindre1993 committed:
Created display comments

Posts now display comments

pretzL committed:
Changed default images to Discord hosting instead of imgur

pretzL committed:
Remove console.logs

pretzL committed:
Fixed comments section not displaying more than one comment at a time, but also limit to 3 to avoid very long posts.

pretzL committed:
Removed excess import

pretzL committed:
"My Profile" buttons now redirect correctly

pretzL committed:
Removed regEx from image handlers since the API now performs this authentication

pretzL committed:
For some reason Bootstrap files disappeared into main branch? Brought them back

pretzL committed:
Added platform for editing profile

pretzL committed:
Delete node_modules directory

pretzL committed:
Trying to add eventlisteners, no dice yet. Will continue later

pretzL committed:
Merge pull request #61 from pretzL/commentSection

Comment section

Sindre1993 committed:
Added profile view on names

Added profile view on names and cleaned up some hardcoded HTML

pretzL committed:
Edit profile media nearly complete, for now it returns error "body must be object"

pretzL committed:
Edit profile media now works

pretzL committed:
Hide follow button when viewing personal profile

pretzL committed:
Merge branch 'js2' into edit-profile

pretzL committed:
Following system initial set up.
Clicking "Follow" on a user adds to localStorage successfully.

pretzL committed:
Side panel is now updated with API info. Not sure why it always returns one undefined entry...

pretzL committed:
Fixed profile side panel styling.
Added comments to filters.

pretzL committed:
Both follow and unfollow now work. Might need a cleanup later though

Sindre1993 committed:
Editting buttons

Some final styling for edit buttons, yet to finish

Sindre1993 committed:
Create single post view

Sindre1993 committed:
updated single post

Sindre1993 committed:
Updated getPost specific

Viewing a single post on own html is now possible by clicking "View Comments", displaying all comments the post has, comments displayed on the home page is limited to 3.

Sindre1993 committed:
Merge pull request #64 from pretzL/edit-profile

Edit profile

Sindre1993 committed:
Merge branch 'js2' into follow-unfollow

Sindre1993 committed:
Merge pull request #67 from pretzL/follow-unfollow

Follow unfollow

pretzL committed:
Merge branch 'js2' into display-comment-specific

pretzL committed:
Merge pull request #68 from pretzL/display-comment-specific

Display comment specific

pretzL committed:
Cleanup

pretzL committed:
Fixed issue with following list returning one index of undefined

Sindre1993 committed:
Created search function

Search now returns filtered api results.

Fixed viewheight on body

Sindre1993 committed:
Fixed some profile styling

Added some styling to profile

pretzL committed:
Merge pull request #69 from pretzL/search-function

Search function

pretzL committed:
Filter on following now works

Sindre1993 committed:
Create post comment

Sindre1993 committed:
Merge pull request #70 from pretzL/filter-by-following

Filter on following now works

pretzL committed:
Merge branch 'js2' into create-comment

Sindre1993 committed:
Create Comments Function

Adding comments to post now working as intended.

pretzL committed:
Merge pull request #71 from pretzL/create-comment

Create comment

Sindre1993 committed:
Made plan for work

Sindre1993 committed:
Started on reactions

created createReaction.mjs, added amount of reactions next to like button.

Sindre1993 committed:
Fixed postComments

It's now possible to post comments on every post.

Sindre1993 committed:
Created Like function

Like button now working

pretzL committed:
When creating a post, tags are now split at "," and are added as an array to the POST request.

pretzL committed:
Added comment

pretzL committed:
Added skeleton loading to API returned elements

pretzL committed:
Fixed minor display errors

pretzL committed:
Merge pull request #75 from pretzL/create-reaction

Create reaction

Sindre1993 committed:
Fixed reaction counter

Reaction counter is now displaying the correct amount of likes, instead of reaction objects.

pretzL committed:
Filter Following was removed somehow... Readded now.

Sindre1993 committed:
Merge pull request #73 from pretzL/split-tags

Split tags into array

Sindre1993 committed:
Removed some unused imports and scripts

Sindre1993 committed:
Merge branch 'js2' into skeleton-loading

Sindre1993 committed:
Merge pull request #74 from pretzL/skeleton-loading

Skeleton loading

pretzL committed:
Fixed issue with "Unexpected end of JSON input" in followFilter

pretzL committed:
Fixed postEdit form responding "post creation successful"

pretzL committed:
Actually fixed issue with followFilter not working

Moesera commented:
# Peer Review

## JS / Functionality
### Landing page / profile page
The profile landing page looks nice, you get a feed of all the posts.
on the profile page the form doesnt work and it only has body, but i guess it is intended.

**What works fine.**
- posting comments works fine.
- View a post's comments looks nice.
- liking post in view post works fine.
- updating banner works great.

**What does not work fine**

- Would enjoy a button to bring me back from view post comments.
- The error message for the avatar update is bit low, couldnt se it when i was editing until i went further down.
- Honestly i dont know what the clear button should do, i expect it to clear the input but doesnt, would just remove it for now if it should not be usable tbh.

### Homepage

**What works fine.**
- editing a post works fine, and i really like the setup
- creating post works fine, other than what i stated down below.
- deleting post works great
- i love your sorting functionality.
- search functionality works fine.

**What does not work fine**
- When i made a post, i didnt get the title back, i only got the body image and tags and can only view body image and tags from others.

### in general
The site functions well and i like alot of the functionality, the code looks good, and the folder structure is well structured in my opinion.

What i am missing is more JsDocs documentation, it is easier to hover over functions and see what gets passed in or what they do and how they work, its easier to manage the code and understand it faster with it. Would recommend to take a look at it, it is also a criteria.
Also when you come back to the code a bit later its easier to get back into your code.

## HTML / Look

- The avatar image scales out of place on the profile page.
- The most buttons i hover over doesnt become a pointer, i would go over the buttons just because it take's away the feel of the website.
- You could also take away the login button when people are logged in at home and my profile page, just doesnt feel right to have it there.
- On smaller screens the hamburger menu does not work.. and you cant navigate because of this.

## in general

The site looks good, and there is not alot of issues. But i would definetly look at the size of the login background image it is 2 MB, you should try to get it down to 200kb max 300.
The image placeholder banner is a bit large too and is around 4000kb.

**The issue:** 
This will cause slow loading speed.

For the image issue i recommend to use the **image** tab then **image size** or **bildestørrelse** in photoshop and shrink it a bit.
And then use the export to web and take down the quality if needed.


### Overall the site looks good, and good luck on the assignment !.

pretzL committed:
Added skeleton loading to search page and fixed an issue where the value wasn't being trimmed to lower case

Sindre1993 committed:
Merge pull request #80 from pretzL/fixing

Fixing

Sindre1993 committed:
Fixed clear button on post creation

"Clear" button will not correctly clear all the inputs.

pretzL committed:
Edit post form now correctly accounts for if there is post media or not.
Created a more sanitized and DRY way of getting the required post info and creating inputs for the form

pretzL committed:
Merge pull request #89 from pretzL/clear-input-fix

Fixed clear button on post creation

Fermain commented:
<img width="839" alt="image" src="https://user-images.githubusercontent.com/10801540/195264072-32f9e76a-ddfb-4fe8-a213-e565b1a4f916.png">

Unable to edit post if there is no media present due to validator

Fermain committed:
Makes this file more modular and resusable

Fermain committed:
Adds error handling

Fermain committed:
Improves format by making use of await

Fermain committed:
Refactor

Fermain commented on src/js/posts/createComment.mjs:
```diff
@@ -0,0 +1,42 @@
+      } else {
+        errorContainer.innerHTML = successMessage("Post creation");
+
+        setTimeout(() => {
```
This is a sign that async/await is not being used properly. Check the `teacher-review` branch for my solution to this.

Fermain commented on src/js/posts/createPost.mjs:
```diff
@@ -0,0 +1,53 @@
+      } else {
+        errorContainer.innerHTML = successMessage("Post creation");
+
+        setTimeout(() => {
```
Here too

Fermain commented on src/js/posts/createReactions.mjs:
```diff
@@ -0,0 +1,31 @@
+      } else {
+        errorContainer.innerHTML = successMessage("Like");
+
+        setTimeout(() => {
```
And here

Fermain commented on src/js/posts/editPost.mjs:
```diff
@@ -0,0 +1,33 @@
+      } else {
+        errorContainer.innerHTML = successMessage("Post edit");
+
+        setTimeout(() => {
```
Here too!

Fermain commented on src/js/posts/getPostToDelete.mjs:
```diff
@@ -0,0 +1,20 @@
+      deletePost(postId);
+
+      // Reloads the website after 1 second after clicking delete.
+      setTimeout(location.reload.bind(location), 1000);
```
You guessed it

Fermain commented on src/js/user/profiles/follow-unfollow/requests.mjs:
```diff
@@ -0,0 +1,36 @@
+      } else {
+        errorContainer.innerHTML = successMessage("Post creation");
+
+        setTimeout(() => {
```
Hello darkness my old friend

Fermain commented on src/js/user/profiles/updateProfile.mjs:
```diff
@@ -0,0 +1,42 @@
+      } else {
+        errorContainer.innerHTML = successMessage("Profile media edit");
+
+        setTimeout(() => {
```
I've come to talk with you again

Fermain commented on src/js/user/registerUser.mjs:
```diff
@@ -0,0 +1,67 @@
+        } else {
+          errorContainer.innerHTML = successMessage("Registration");
+
+          setTimeout(() => {
```
Because a vision softly creeping,
Left its seeds while I was sleeping

Fermain reviewed (changes requested):
This is a really high quality submission that looks and feels like a real application. There are some issues to be resolved (setTimeout mostly) and there are improvements to be made in terms of code organisation and structure. (await > then).

With the setTimeout issue and update post issues solved this is a very, very comfortable pass.

pretzL committed:
Add skeleton loading to search page user panel

Sindre1993 committed:
Removed Login Button

Removed login button from all pages but login.html.

Working as intended

Sindre1993 committed:
post.html updated

posts now contain a back button that takes you back to the previous page.

Tested and working as intended

Sindre1993 committed:
error container fixed

Error container on profile.html moved up for better UX.

Sindre1993 committed:
Merge pull request #90 from pretzL/fix-editPost

Edit post form now correctly accounts for if there is post media or not.

pretzL committed:
Merge pull request #97 from pretzL/login-button-fix

Removed Login Button

pretzL committed:
Merge pull request #98 from pretzL/post-detail-backbutton

post.html updated

Sindre1993 committed:
Added hover state on li.

Listed items in cogwheel now has a pointer hover state.

pretzL committed:
Merge pull request #99 from pretzL/error-container-placement

error container fixed

pretzL committed:
Merge pull request #100 from pretzL/dropdown-hover-fix

Added hover state on li.

Sindre1993 committed:
Merge pull request #102 from pretzL/teacher-review

Teacher review

Sindre1993 committed:
fixed images in posts

Post images are now centered correctly with a 100% width and a fixed height.

Working as intended

Sindre1993 committed:
profile image fixed

Profile image now properly centered when compressing viewport.

Working as intended

pretzL committed:
All fetches refactored in accordance with teacher feedback

Sindre1993 committed:
login background fix

Background image on login page size has now been lowered from 59mb to 350kb.

Loading should now be much shorter.

Sindre1993 committed:
Updated Drop Down mobile

Dropdown menu on mobile now working, added class to center the dropdown as it would clip through the sides below 600px.

Working as intended

Sindre1993 committed:
post-social-icons fixed

Post-social-icons is now wrapping instead of breaking out of container.

Working as intended

Sindre1993 committed:
profile.html post fix

profile.html now has the correct post template.

Working as intended

pretzL committed:
Merge branch 'js2' into post-media-imagefix

pretzL committed:
Merge pull request #103 from pretzL/post-media-imagefix

fixed images in posts

pretzL committed:
Merge branch 'js2' into profile-image-fix

pretzL committed:
Merge pull request #104 from pretzL/profile-image-fix

profile image fixed

pretzL committed:
Merge branch 'js2' into login-image-fix

pretzL committed:
Merge pull request #105 from pretzL/login-image-fix

login background fix

pretzL committed:
Merge branch 'js2' into hamburger-menu-fix

pretzL committed:
Merge pull request #107 from pretzL/hamburger-menu-fix

Updated Drop Down mobile

pretzL committed:
Merge pull request #109 from pretzL/button-post-fix

post-social-icons fixed

pretzL committed:
Merge pull request #110 from pretzL/profile-postcontainer-fix

profile.html post fix

pretzL committed:
Merge branch 'js2' into filter-feedback

pretzL committed:
Merge pull request #111 from pretzL/filter-feedback

Filter feedback fix

Sindre1993 committed:
Filter feedback fix

Filters now change innerText of button for feedback on current filter.

pretzL committed:
Fixed issue with editing a post returning "body/media must match format uri" error

pretzL committed:
Added new form of timed redirect

Sindre1993 committed:
Merge pull request #106 from pretzL/refactor-fetches

All fetches refactored in accordance with teacher feedback

Sindre1993 committed:
Custom scrollbar added

Custom scrollbar added for certain browsers.

pretzL committed:
Added JSDocs for code up until profileFormTemplate

pretzL committed:
Created remaining JSDocs
Changed usage of getUserAuth() to getItem()
Fixed fetch format for another few fetches

pretzL committed:
Added success message and timeout to login

pretzL committed:
Generated documentation for JSDocs

pretzL committed:
Update README.md

Sindre1993 committed:
Merge pull request #113 from pretzL/jsdocs

Added JSDocs

Sindre1993 committed:
Updated Custom Scrollwheel

pretzL committed:
Compiled SASS

pretzL committed:
Merge pull request #112 from pretzL/custom-scrollbar

Custom scrollbar added

pretzL committed:
Fixed issue with register not working properly due to missing imports.
Fixed nav-active class on profile page being incorrect

Fermain committed:
Fixes import errors

Fermain committed:
Adds missing import

pretzL committed:
Merge pull request #114 from pretzL/js2-marking

Js2 marking

pretzL committed:
Merge branch 'js2' into fixing

pretzL committed:
Merge pull request #115 from pretzL/fixing

Fixed issue with register not working properly due to missing imports.

pretzL committed:
Add missing import

pretzL committed:
Fixed issue with searchbar not working due to double searchbar

pretzL committed:
Fixed issue with profile displaying other posts than users' own

Sindre1993 committed:
Fixed Home Button & Searchbar on mobile view

Fixed home button, the entire icon in addition to the text is now clickable and will redirect you to the home page.

Searchbar now has a width that doesn't clip through the screen on mobile view.

Tested for 10 minutes on all pages, working as intended

Sindre1993 committed:
Merge pull request #116 from pretzL/fixes

Fixes

Fermain reviewed (changes requested):
It looks like you have some artefacts left over from installing a bundler (out directory).

If these files aren't being used they should be removed from history. It's conventional when using a bundler to ignore the output directory so that this content is not included in a pull request.

Otherwise, the code is working great.

pretzL committed:
Untrack "out" folder from Vite

pretzL committed:
Delete out directory
